crates/subspace-networking/src/protocols/connected_peers/tests.rs

```diff
@@ -0,0 +1,272 @@
+            dialing_interval,
+            ..Default::default()
+        }),
+    );
+
+    peer1.listen().await;
+
+    loop {
+        select! {
+            event = peer1.next_swarm_event().fuse() => {
+                if matches!(event, SwarmEvent::Behaviour(ConnectedPeersEvent::<ConnectedPeersInstance>::NewDialingCandidatesRequested(..))){
+                    break;
+                }
+            },
+            _ = sleep(large_delay).fuse() => {
+                panic!("No new peers requests.");
+            }
+        }
+    }
+}
+
+#[tokio::test()]
+async fn test_target_connected_peer_limit_number() {
+    let decision_timeout = Duration::from_millis(500);
+    let large_delay = Duration::from_millis(2000);
+    let target_connected_peers = 1;
+
+    let mut peer1 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            target_connected_peers,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer2 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            target_connected_peers,
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer3 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            target_connected_peers,
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    peer2.listen().await;
+    peer3.listen().await;
+
+    peer1.connect(&mut peer2).await;
+    peer1.connect(&mut peer3).await;
+
+    peer2.connect(&mut peer3).await;
+
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer3.local_peer_id(), true);
+
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), true);
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer3.local_peer_id(), true);
+
+    peer3
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), true);
+    peer3
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+
+    loop {
+        select! {
+            _ = peer1.next_swarm_event().fuse() => {},
+            _ = peer2.next_swarm_event().fuse() => {},
+            _ = peer3.next_swarm_event().fuse() => {},
+            _ = sleep(large_delay).fuse() => {
+                break;
+            }
+        }
+    }
+
+    assert!(peer1.is_connected(peer2.local_peer_id()));
+    assert!(!peer1.is_connected(peer3.local_peer_id()));
+
+    assert!(peer2.is_connected(peer1.local_peer_id()));
+    assert!(!peer2.is_connected(peer3.local_peer_id()));
+
+    assert!(!peer3.is_connected(peer1.local_peer_id()));
+    assert!(!peer2.is_connected(peer2.local_peer_id()));
```

Comment:
I don't think I understand why this is the result. For instance why wouldn't the first peer be connected to the third peer.

Reply:
Connection slots are filled on peer1 and peer2 because they got connected to each other earlier and `target_connected_peers = 1`.

Reply:
That is the kind of explanation I was expecting to see in comments of the test.

What is `assert!(!peer2.is_connected(peer2.local_peer_id()));` supposed to check, that peer is not connected to itself?